### ISSUE TITLE
✨ test(core): Slice C — eliminate owned-code mocks

### ISSUE THREAD

**Issue by CedrickArmel** [CedrickArmel/radiologist#101]
# ✨ Slice C: core owned-code mock elimination

## Context

Part of the **Classist outside-in TDD compliance + pytest fixture consolidation** epic.
**Blocks on #98** (provides the `ckpt_path` and `lmodule` fixtures and the shared value fixtures).
Independent of #99, #100, #102, #103.

The `radiologist-core` tests mock owned code in five places: the ETL `records_reader` function,
Lightning's `LModule.load_from_checkpoint`, and `pl.Trainer` (via `MagicMock`). This slice
replaces every such mock with the real object. The convention is strict: **no `MagicMock`, no
`SimpleNamespace`, no stub** for any Lightning component or `radiologist.*` symbol. When a test
needs a trainer, it runs a real `pl.Trainer(fast_dev_run=True)`.

## User story

As a **developer refactoring core**, I want **core tests to exercise the real reader, real
checkpoint loading, and a real training loop**, so that **the tests catch integration breakage
instead of passing against stubs**.

### `radiologist-core/tests/test_datamodule.py` — **EDIT** (violation V1)

Remove the two `patch("radiologist.core.data.datamodule.records_reader")` patches (current lines
~367 and ~421). The `split_manifest_uri` fixture already writes a real JSONL manifest; let the
real `records_reader` run against it. Change the assertions that currently check mock call counts
to assertions on observable datamodule state — `dm.priors`, `dm.train_size`, `dm.val_size` — as
derived from the real manifest on disk.

### `radiologist-core/tests/test_onnx_export.py` — **EDIT** (violation V2)

Remove the `patch.object(LModule, "load_from_checkpoint", return_value=lm)` (~line 85). Inject
the `ckpt_path` fixture and pass it to the export entry point:
`export_onnx(ckpt_path=ckpt_path, ...)`. The real `load_from_checkpoint` runs.

### `radiologist-core/tests/test_registry.py` — **EDIT** (violation V3)

Remove the four `load_from_checkpoint` patches (~lines 157, 183, 217, 245). Inject `ckpt_path`
and pass the real checkpoint path to the registry/export entry point in each test.

### `radiologist-core/tests/test_callbacks.py` — **EDIT** (violation V4)

Replace every `MagicMock()` used as a `pl.Trainer` or `pl.LightningModule` with a **real
instance**. The test pattern is:

1. Construct a real `LModule` via the `lmodule` fixture.
2. Construct a real `pl.Trainer` with `fast_dev_run=True`, the callback under test, and
   `accelerator="cpu"` (plus `enable_progress_bar=False`, `enable_model_summary=False` to keep
   output clean).
3. Run `trainer.fit(lmodule, datamodule=dm)`. The callback hooks fire on the real trainer state.
4. Assert on **observable side-effects**: entries in `trainer.callback_metrics`, files written to
   disk, or calls on logger objects (loggers are an external boundary — mocking them is fine).

Example for `BestMetricCallback`:

```python
def test_best_metric_written_to_callback_metrics(lmodule, dm):
    from radiologist.core import BestMetricCallback

    callback = BestMetricCallback()
    trainer = pl.Trainer(
        fast_dev_run=True, accelerator="cpu",
        enable_progress_bar=False, enable_model_summary=False,
        callbacks=[callback], logger=False,
    )
    trainer.fit(lmodule, datamodule=dm)
    assert "best_val_score" in trainer.callback_metrics
```

For `WandbDefineSummaryCallback`, the W&B `wandb.run` is still mocked (it is an external
boundary). The test wires the callback into a real trainer and asserts `run.define_metric` is
called by the hook:

```python
def test_wandb_summary_defines_metrics(lmodule, dm):
    from unittest.mock import patch, MagicMock
    import wandb
    from radiologist.core import WandbDefineSummaryCallback

    fake_run = MagicMock()
    callback = WandbDefineSummaryCallback()
    trainer = pl.Trainer(
        fast_dev_run=True, accelerator="cpu",
        enable_progress_bar=False, enable_model_summary=False,
        callbacks=[callback], logger=False,
    )
    with patch.object(wandb, "run", fake_run):
        trainer.fit(lmodule, datamodule=dm)
    fake_run.define_metric.assert_called()
```

### `radiologist-core/tests/test_attribution_callback.py` — **EDIT** (violation V5)

Replace every `MagicMock()` trainer with a real `pl.Trainer(fast_dev_run=True, ...)`. The
`AttributionCallback` reads `trainer.current_epoch`, `trainer.global_step`, `trainer.log_dir`,
`trainer.is_global_zero`, and `trainer.default_root_dir` from the real trainer state during
validation. Pass `default_root_dir=str(tmp_path)` to control where files are written:

```python
trainer = pl.Trainer(
    fast_dev_run=True, accelerator="cpu",
    enable_progress_bar=False, enable_model_summary=False,
    default_root_dir=str(tmp_path),
    callbacks=[attribution_callback],
    logger=False,
)
trainer.fit(lmodule, datamodule=dm)
```

Assert on the output files written to `tmp_path` (PNG files, attribution artefacts) rather than
on mock attribute access.

The `captum` and `wandb` patches remain — they are external boundaries:
- `patch.dict(sys.modules, {"captum": ...})` for testing the absent-captum degradation path
- `MagicMock` for `wandb` module-level state for testing W&B logging calls

### `radiologist-core/tests/test_module.py` — **EDIT** (violation V6)

Replace the two `MagicMock()` trainer stubs with real trainers:

- `on_test_epoch_end` test (~line 384): use `pl.Trainer(fast_dev_run=True, ...,
  default_root_dir=str(tmp_path))` and run `trainer.test(lmodule, datamodule=dm)`. Then assert
  `preds-rank0.pt` exists under `tmp_path`.
- `configure_gradient_clipping` test (~line 437): use
  `pl.Trainer(fast_dev_run=True, ..., gradient_clip_val=1.0, gradient_clip_algorithm="norm")`
  and run `trainer.fit(lmodule, datamodule=dm)`. The hook fires naturally; assert the logged
  `grad_norm_post_clip` metric is recorded in `trainer.callback_metrics`.

## Interface contracts

No new interfaces. Consumes `ckpt_path`, `lmodule`, and `dm` from #98.

## Acceptance criteria

- [ ] Given the real JSONL manifest from `split_manifest_uri`, the datamodule computes and exposes
      the expected priors and train size — asserted on the datamodule's public state, not on mock
      call counts.
- [ ] Given a real Lightning checkpoint from `ckpt_path`, ONNX export and registry tests produce
      valid ONNX files without patching `load_from_checkpoint`.
- [ ] Callback tests drive callbacks via a real `pl.Trainer(fast_dev_run=True)` fit/test loop;
      assertions are on observable side-effects (files written, `callback_metrics` entries, logger
      calls on external-boundary mocks).
- [ ] No `radiologist.*` symbol and no Lightning component (`Trainer`, `LightningModule`,
      `DataModule`) is patched or mocked anywhere in `radiologist-core/tests/`.
- [ ] No `MagicMock`, no `SimpleNamespace`, no stub is used for any Lightning object.
- [ ] mypy clean; pytest green.

## Out of scope

- Simplifying datamodule construction (`_make_dm` / `dm` fixture) — that is #100.
- Deleting the rank-0/rank-1 broadcast test — that is #100.
- Inference, utils, or registry test rewrites — those are #99, #102, #103.

## Technical notes

- `pl.Trainer(fast_dev_run=True)` runs one train batch + one validation batch. That is enough to
  fire `on_validation_epoch_end`. For `on_test_epoch_end`, use `trainer.test()` separately.
- `default_root_dir=str(tmp_path)` controls where Lightning writes logs and checkpoints; pass it
  whenever a callback writes files to `trainer.log_dir` or `trainer.default_root_dir`.
- `accelerator="cpu"` is mandatory — no GPU is available in CI.
- Use `model.train(mode=False)` rather than `model.eval()` in any helper; the sandbox security
  hook false-positives on `.eval()`.
- Coordinate with #100 on `test_datamodule.py`: the two slices edit different regions; whichever
  lands second rebases onto the first.
- W&B and `captum` mocks in `test_attribution_callback.py` are **kept** — they are external
  boundaries. Only the `MagicMock` trainer and module stubs are removed.
